### PR TITLE
[Synapse] Make Synapse PowerShell consume track 2 Spark SDK

### PR DESCRIPTION
**Pull request by wonner**
## Description

Make Synapse PowerShell consume track 2 Spark SDK.

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [ ] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

**Comment by adxsdkps** [comment]
Can one of the admins verify this patch?

**Comment by idear1203** [comment]
@VeryEarly @erich-wang We meet the following CI build error:

<pre>
  Analyzer invoked with parameters: -p D:\a\1\s\artifacts/Debug -r D:\a\1\s\artifacts//StaticAnalysisResults -h -u -m
  No value provided for the --modules-to-analyze parameter. Filtering over all built modules.
  <b>Analyzing module under D:\a\1\s\artifacts\Debug\Az.Accounts ...
  Analyzing module under D:\a\1\s\artifacts\Debug\Az.Storage ...
  Help record Set-AzDataLakeGen2ItemACLObject has no cmdlet.
  Analyzing module under D:\a\1\s\artifacts\Debug\Az.Synapse ...</b>
  
  Unhandled Exception: System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types.
  Could not load file or assembly 'Azure.Core, Version=1.2.2.0, Culture=neutral, PublicKeyToken=92742159e12e44c8'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621)
  Could not load file or assembly 'Azure.Core, Version=1.2.2.0, Culture=neutral, PublicKeyToken=92742159e12e44c8'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621)
  Could not load file or assembly 'Azure.Core, Version=1.2.2.0, Culture=neutral, PublicKeyToken=92742159e12e44c8'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621)
     at Tools.Common.Loaders.CmdletLoader.GetModuleMetadata(String assemblyPath) in D:\a\1\s\tools\Tools.Common\Loaders\CmdletLoader.cs:line 209
     at Tools.Common.Loaders.CmdletLoader.GetModuleMetadata(String assemblyPath, List`1 commonOutputFolders) in D:\a\1\s\tools\Tools.Common\Loaders\CmdletLoader.cs:line 54
     at StaticAnalysis.HelpAnalyzer.HelpAnalyzer.AnalyzeMarkdownHelp(IEnumerable`1 scopes, String directory, ReportLogger`1 helpLogger, List`1 processedHelpFiles, String savedDirectory) in D:\a\1\s\tools\StaticAnalysis\HelpAnalyzer\HelpAnalyzer.cs:line 279
     at StaticAnalysis.HelpAnalyzer.HelpAnalyzer.Analyze(IEnumerable`1 scopes, IEnumerable`1 modulesToAnalyze) in D:\a\1\s\tools\StaticAnalysis\HelpAnalyzer\HelpAnalyzer.cs:line 111
     at StaticAnalysis.Program.Main(String[] args) in D:\a\1\s\tools\StaticAnalysis\Program.cs:line 139
D:\a\1\s\build.proj(208,5): error MSB3073: The command "dotnet D:\a\1\s\artifacts/StaticAnalysis/StaticAnalysis.Netcore.dll -p D:\a\1\s\artifacts/Debug -r D:\a\1\s\artifacts//StaticAnalysisResults -h -u -m " exited with code -532462766.
##[error]Error: The process 'C:\Program Files\dotnet\dotnet.exe' failed with exit code 1
</pre>

We did some investigation locally. It seems that there are some issues when both `Az.Synapse` and `Az.Storage` exist under the `artifacts\Debug` folder.

If I remove the `Az.Storage` from `artifacts\Debug` folder, the command works fine:

<pre>
PS> dotnet D:\code\AzureSDK\azure-powershell\artifacts\StaticAnalysis\StaticAnalysis.Netcore.dll -p D:\code\AzureSDK\azure-powershell\artifacts\Debug -r D:\code\AzureSDK\azure-powershell\artifacts\StaticAnalysisResults -h -u -m
Analyzer invoked with parameters: -p D:\code\AzureSDK\azure-powershell\artifacts\Debug -r D:\code\AzureSDK\azure-powershell\artifacts\StaticAnalysisResults -h -u -m
No value provided for the --modules-to-analyze parameter. Filtering over all built modules.
<b>Analyzing module under D:\code\AzureSDK\azure-powershell\artifacts\Debug\Az.Accounts ...
Analyzing module under D:\code\AzureSDK\azure-powershell\artifacts\Debug\Az.Synapse ...</b>
PS>
</pre>

But the command will fail if `Az.Storage` exist.

<pre>
PS> dotnet D:\code\AzureSDK\azure-powershell\artifacts\StaticAnalysis\StaticAnalysis.Netcore.dll -p D:\code\AzureSDK\azure-powershell\artifacts\Debug -r D:\code\AzureSDK\azure-powershell\artifacts\StaticAnalysisResults -h -u -m
Analyzer invoked with parameters: -p D:\code\AzureSDK\azure-powershell\artifacts\Debug -r D:\code\AzureSDK\azure-powershell\artifacts\StaticAnalysisResults -h -u -m
No value provided for the --modules-to-analyze parameter. Filtering over all built modules.
<b>Analyzing module under D:\code\AzureSDK\azure-powershell\artifacts\Debug\Az.Accounts ...
Analyzing module under D:\code\AzureSDK\azure-powershell\artifacts\Debug\Az.Storage ...
Help record Set-AzDataLakeGen2ItemACLObject has no cmdlet.
Analyzing module under D:\code\AzureSDK\azure-powershell\artifacts\Debug\Az.Synapse ...</b>

Unhandled Exception: System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types.
Could not load file or assembly 'Azure.Core, Version=1.2.2.0, Culture=neutral, PublicKeyToken=92742159e12e44c8'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621)
Could not load file or assembly 'Azure.Core, Version=1.2.2.0, Culture=neutral, PublicKeyToken=92742159e12e44c8'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621)
Could not load file or assembly 'Azure.Core, Version=1.2.2.0, Culture=neutral, PublicKeyToken=92742159e12e44c8'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621)
   at StaticAnalysis.Program.Main(String[] args) in D:\code\AzureSDK\azure-powershell\tools\StaticAnalysis\Program.cs:line 149
</pre>

Is it because the Synapse project references the Storage project? Could you please provide suggestions about how we should fix the issue? Thanks in advance.

File: src/Synapse/Synapse/Synapse.csproj [hunk 1]
```diff
@@ -30,4 +30,9 @@
     </EmbeddedResource>
   </ItemGroup>
 
+	<Target Name="CopyFiles" AfterTargets="Build">
+		<Copy SourceFiles="@(PreLoadAssemblies)" DestinationFolder="$(TargetDir)\PreloadAssemblies" />
+		<Copy SourceFiles="@(NetCoreAssemblies)" DestinationFolder="$(TargetDir)\NetCoreAssemblies" />
+	</Target>
+
```
Comment (erich-wang) [inline src/Synapse/Synapse/Synapse.csproj:37]: Please remove these lines

File: src/Synapse/Synapse/Common/AzureSessionCredential.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            if (environment != null)
+            {
+                if (!environment.IsPropertySet(SynapseOAuthEndpointResourceKey))
+                {
+                    environment.SetProperty(SynapseOAuthEndpointResourceKey, SynapseOAuthEndpointResourceValue);
+                }
+            }
+
+            return environment;
+        }
```
Comment (erich-wang) [inline src/Synapse/Synapse/Common/AzureSessionCredential.cs:78]: The environment related change should be in another repo https://github.com/Azure/azure-powershell-common, you may search related code for https://github.com/Azure/azure-powershell-common/blob/master/src/Authentication.Abstractions/AzureEnvironmentConstants.cs#L118..L121 as reference.
BTW, does synapse have same endpoint for different azure cloud environments? E.g. UsGov, GermanGov environments.

Comment (idear1203) [reply src/Synapse/Synapse/Common/AzureSessionCredential.cs:78]: @erich-wang , we already added environment related config in azure-powershell-common: https://github.com/Azure/azure-powershell-common/blob/master/src/Authentication.Abstractions/AzureEnvironmentConstants.cs#L194-L195 . We added Synapse related endpoints as extended properties. But this comment makes sense to me. @wonner , could you please try removing this method and authenticate as below (line 27)?

``` csharp
            IAccessToken accessToken1 = AzureSession.Instance.AuthenticationFactory.Authenticate(
               DefaultContext.Account,
               DefaultContext.Environment,
               DefaultContext.Tenant.Id,
               null,
               ShowDialog.Never,
               null,
               AzureEnvironment.ExtendedEndpoint.AzureSynapseAnalyticsEndpointResourceId);
```

Comment (idear1203) [reply src/Synapse/Synapse/Common/AzureSessionCredential.cs:78]: @erich-wang for your 2nd question. Currently Synapse only have endpoints for Global cloud and China cloud. Other national clouds are not supported.

File: src/Synapse/Synapse/Common/AzureSessionCredential.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+        /// <summary>
+        /// The extension key to use for the synapse token audience value
+        /// </summary>
+        public const string SynapseOAuthEndpointResourceKey = "SynapseOAuthEndpointResourceKey";
+
+        /// <summary>
+        /// Default resourceId for synapse OAuth tokens
+        /// </summary>
+        public const string SynapseOAuthEndpointResourceValue = "https://dev.azuresynapse.net";
```
Comment (idear1203) [inline src/Synapse/Synapse/Common/AzureSessionCredential.cs:75]: Do we still need these two constants?

File: src/Synapse/Synapse/Synapse.csproj [hunk 4]
```diff
@@ -12,7 +12,7 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.Azure.Management.Synapse" Version="0.1.0-preview.2" />
-    <PackageReference Include="Microsoft.Azure.Synapse" Version="0.1.0-preview" />
+	<PackageReference Include="Azure.Analytics.Synapse.Spark" Version="1.0.0-preview.1" />
```
Comment (idear1203) [inline src/Synapse/Synapse/Synapse.csproj:15]: tab to spaces?

File: src/Synapse/Synapse/Az.Synapse.psd1 [hunk 5]
```diff
@@ -57,7 +57,7 @@ RequiredModules = @(@{ModuleName = 'Az.Accounts'; ModuleVersion = '1.9.0'; })
 
 # Assemblies that must be loaded prior to importing this module
 RequiredAssemblies = 'Microsoft.Azure.Management.Synapse.dll', 
-               'Microsoft.Azure.Synapse.dll'
```
Comment (idear1203) [inline src/Synapse/Synapse/Az.Synapse.psd1:60]: Remove 'Microsoft.Azure.Synapse.dll'?

**Pull request merged**